sql: INVERTED JOIN acceleration for `json ? string`, `json ?& array`, `json ?| array` and `array && array`

ZhouXing19 opened the pull request:
fixes https://github.com/cockroachdb/cockroach/issues/81642
fixes https://github.com/cockroachdb/cockroach/issues/81643

ref: https://github.com/cockroachdb/cockroach/pull/81253

This commit adds support for accelerating the json ? string, json ?& array, json ?| array and array && array operators for INVERTED JOIN when the json/array argument is a column that has an inverted index over it.

Release note (sql change): the json ? string, json ?& array, json ?| array and array && array operators are now index accelerated for INVERTED JOIN statement if there is an inverted index over the json column referred to on the left hand side of the expression.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/149898)
<!-- Reviewable:end -->

yuzefovich reviewed (commented):
Nice! I have some minor questions and suggestions.

nit: if you want the PR to close multiple issues, then every issue number needs to be prefixed with "fixes" or a similar keyword ("fixes: issue1, issue2" will only close issue1 automatically).

> Reviewed 1 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/149898#-:-OV9DZfiBKDdnjMBlbfY:b-9grs8k)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @michae2, and @rytaft)

___
*[`pkg/sql/logictest/testdata/logic_test/inverted_join_json_array` line 65 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/149898#-OV9G47D49n5__wL2Hcq:-OV9G47D49n5__wL2Hcr:b5brk70) ([raw file](https://github.com/cockroachdb/cockroach/blob/805621d35531674bc9d553220f89c27272aebff4/pkg/sql/logictest/testdata/logic_test/inverted_join_json_array#L65)):*
> ```
>   (43, '[{"a": "b", "c": "d"}]'),
>   (44, '[{"a": "b"}, {"c": "d"}]'),
>   (45, '{"a": "a"}'),
> ```

nit: it'd be cleaner to add these two new rows into the test table in a separate commit (which would also contain the updates to all existing tests), so that the main commit only had test changes that correspond to the new functionality.
___
*[`pkg/sql/opt/invertedidx/json_array.go` line 36 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/149898#-OV9DCYi8GR4hT7oykBS:-OV9DCYi8GR4hT7oykBT:b6hc0rh) ([raw file](https://github.com/cockroachdb/cockroach/blob/805621d35531674bc9d553220f89c27272aebff4/pkg/sql/opt/invertedidx/json_array.go#L36)):*
> ```Go
> }
> 
> // Extend it with memo.JsonExistsExpr
> ```

nit: seems like a leftover?
___
*[`pkg/sql/opt/invertedidx/json_array.go` line 123 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/149898#-OV9Ec3Z8sw3gfrNWgDf:-OV9Ec3Z8sw3gfrNWgDg:bc4d4mg) ([raw file](https://github.com/cockroachdb/cockroach/blob/805621d35531674bc9d553220f89c27272aebff4/pkg/sql/opt/invertedidx/json_array.go#L123)):*
> ```Go
> 	// If commuteArgs is true, we construct a new equivalent expression so that
> 	// the left argument is the indexed column.
> 	if commuteArgs {
> ```

Hm, do we not need to make a new expression here? Is it ever possible that we'll have `commuteArgs=true` for one of the newly supported expressions? If not, then maybe we should replace the `if` block with the type switch where we'd return an assertion error when trying to commute one of the new expressions?
___
*[`pkg/sql/opt/invertedidx/json_array.go` line 321 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/149898#-OV9FggoBkyLpq_vk9yY:-OV9FggoBkyLpq_vk9yZ:btzz6b) ([raw file](https://github.com/cockroachdb/cockroach/blob/805621d35531674bc9d553220f89c27272aebff4/pkg/sql/opt/invertedidx/json_array.go#L321)):*
> ```Go
> 				return nil, nil
> 			}
> 			switch t.Operator.Symbol {
> ```

nit: this type switch is now a copy of the one in `NewJSONOrArrayDatumsToInvertedExpr` - should we extract a shared helper function for both? Up to you - I don't have a strong opinion here.


<!-- Sent from Reviewable.io -->

ZhouXing19 committed:
logictest: add more entries in the json_tab in inverted_join_json_array

Release note: None

ZhouXing19 committed:
sql: INVERTED JOIN acceleration for json {?, ?&, ?|} array and array && array

fixes: #81642
fixes: #81643

This commit adds support for accelerating the json ? string,
json ?& array, json ?| array and array && array operators
for INVERTED JOIN when the json/array argument is a column that has an
inverted index over it.

Release note (sql change): the json ? string, json ?& array,
json ?| array and array && array operators are now index
accelerated for INVERTED JOIN statement if there is an
inverted index over the json column referred
to on the left hand side of the expression.

ZhouXing19 commented:
Ohh cool thanks!





<!-- Sent from Reviewable.io -->

ZhouXing19 commented:
*[`pkg/sql/opt/invertedidx/json_array.go` line 321 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/149898#-OV9FggoBkyLpq_vk9yY:-OVEYfN49qrQzoUO-mvr:bz5py0o) ([raw file](https://github.com/cockroachdb/cockroach/blob/805621d35531674bc9d553220f89c27272aebff4/pkg/sql/opt/invertedidx/json_array.go#L321)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: this type switch is now a copy of the one in `NewJSONOrArrayDatumsToInvertedExpr` - should we extract a shared helper function for both? Up to you - I don't have a strong opinion here.
</blockquote></details>

Good idea! Done.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (approved):
Nice work! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 3 of 3 files at r2, 3 of 3 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/149898#-:-OVGREbJ3BFZS0rtwcke:b-8ewgme)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @mgartner, @michae2, @rytaft, and @ZhouXing19)

___
*[`pkg/sql/opt/invertedidx/json_array.go` line 123 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/149898#-OV9Ec3Z8sw3gfrNWgDf:-OVGH5SGEPljOZDN1yJc:b-rnbwxd) ([raw file](https://github.com/cockroachdb/cockroach/blob/805621d35531674bc9d553220f89c27272aebff4/pkg/sql/opt/invertedidx/json_array.go#L123)):*
<details><summary><i>Previously, ZhouXing19 (Jane Xing) wrote…</i></summary><blockquote>

Good catch, this was a bug that `commuteArgs` could be `true` for `overlaps` too, so previously an `overlaps` could be wrongly converted into a `ContainedBy` expr. Now fixed. 

However, oddly, with the original bug, the logic test results for the `overlaps` operator is correct, i.e. this bug doesn't give a result of `ContainedBy`, but gives the right result of `Overlaps`. E.g. 

```sql

statement ok
CREATE TABLE array_tab (
  a INT PRIMARY KEY,
  b INT[]
)

statement ok
CREATE INVERTED INDEX foo_inv ON array_tab(b)

statement ok
INSERT INTO array_tab VALUES
  (1, '{}'),
  (2, '{1}'),
  (3, '{2}'),
  (4, '{1, 2}'),
  (5, '{1, 3}'),
  (6, '{1, 2, 3, 4}'),
  (7, ARRAY[NULL]::INT[]),
  (8, NULL)

statement ok
CREATE TABLE array_tab_not_idx (
  a INT PRIMARY KEY,
  b INT[]
);

statement ok
INSERT INTO array_tab_not_idx VALUES
  (1, '{}'),
  (2, '{1}'),
  (3, '{2}'),
  (4, '{1, 2}'),
  (5, '{1, 3}'),
  (6, '{1, 2, 5}'),
  (7, ARRAY[NULL]::INT[]),
  (8, NULL)

# Ensure that join confition with not-index-array && index-array will be
# normalized to index-array && not-index-array.
query ITIT
SELECT * FROM array_tab_not_idx@array_tab_not_idx_pkey AS a2
INNER INVERTED JOIN array_tab@foo_inv AS a1
ON a2.b && a1.b AND a2.a > 1 AND a2.a < 7
ORDER BY a1.a, a2.a
----
```

I ran it with the code before fixing this bug you mentioned, and debugger shows me that we did hit `return j.factory.ConstructContainedBy(right, left)`, so we did wrongly converted an `overlap` expr to a `containedby` expr. But the result still contains the row `6  {1,2,5}  6  {1,2,3,4}`, which wouldn't be produced by a `containedby` expr.
</blockquote></details>

Interesting. I couldn't explain it, so I looked a bit closer, and it's due to how we handle ContainedBy in the inverted index - take a look at `rowenc.encodeContainedArrayInvertedIndexSpans`. There, we construct spans to read in the inverted index effectively as "logical OR of all possible single-element sub-arrays" (meaning that if we're trying evaluate `array_col <@ [1, 2, 3]`, we will look up all index entries that have an element `1` or element `2` or element `3`); these spans are "not tight", which means that we have to do additional filtering later to remove rows that don't actually pass the ContainedBy condition (in other words, we might get false positives, so we need to remove them).

I reverted the fix, and the bug can be seen in different EXPLAIN variants, perhaps the easiest is EXPLAIN (OPT):
```
                  │    └── inner-join (lookup array_tab [as=a1])
                  │         ├── lookup columns are key
                  │         ├── inner-join (inverted array_tab@foo_inv,inverted [as=a1])
                  │         │    ├── flags: force inverted join (into right side)
                  │         │    ├── inverted-expr
                  │         │    │    └── a1.b <@ a2.b
                  │         │    ├── scan array_tab_not_idx [as=a2]
                  │         │    │    ├── constraint: /1: [/2 - /6]
                  │         │    │    └── flags: force-index=array_tab_not_idx_pkey
                  │         │    └── filters (true)
                  │         └── filters
                  │              └── a2.b && a1.b
```
Note that the inverted expr is incorrectly replaced with ContainedBy, but then, after we perform lookup join into the PK of `array_tab`, we apply the correct filter (because that filter wasn't replaced in `extractJSONOrArrayJoinCondition`). It just so happened that this bug had no actual effect on the output :smiley: 
___
*[`-- commits` line 9 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/149898#-OVGG75cCG0ns8hdU1VF:-OVGG75cCG0ns8hdU1VG:bj48h54):*
By the way, only the PR description matters for auto-closing issues via "fixes" keyword, and it does nothing when only present in the commit message. (I adjusted the PR description accordingly.)
___
*[`pkg/sql/opt/invertedidx/json_array.go` line 131 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/149898#-OVGGiKKDOCZ-lsVG7lo:-OVGGiKKDOCZ-lsVG7lp:blbjbm7) ([raw file](https://github.com/cockroachdb/cockroach/blob/395ad63438dca8d064d8a8f9bac2768c8f220141/pkg/sql/opt/invertedidx/json_array.go#L131)):*
> ```Go
> 			// commuteArgs should only be true for operators where both sides
> 			// could be json or array columns.
> 			return nil
> ```

nit: let's return an assertion failure in crdb-test builds (i.e. gated behind `buildutil.CrdbTestBuild` boolean) since this seems unexpected.


<!-- Sent from Reviewable.io -->

ZhouXing19 commented:
*[`pkg/sql/opt/invertedidx/json_array.go` line 123 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/149898#-OV9Ec3Z8sw3gfrNWgDf:-OVJZ-CU5Q62Q-zKKgKa:bxp5jsw) ([raw file](https://github.com/cockroachdb/cockroach/blob/805621d35531674bc9d553220f89c27272aebff4/pkg/sql/opt/invertedidx/json_array.go#L123)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Interesting. I couldn't explain it, so I looked a bit closer, and it's due to how we handle ContainedBy in the inverted index - take a look at `rowenc.encodeContainedArrayInvertedIndexSpans`. There, we construct spans to read in the inverted index effectively as "logical OR of all possible single-element sub-arrays" (meaning that if we're trying evaluate `array_col <@ [1, 2, 3]`, we will look up all index entries that have an element `1` or element `2` or element `3`); these spans are "not tight", which means that we have to do additional filtering later to remove rows that don't actually pass the ContainedBy condition (in other words, we might get false positives, so we need to remove them).

I reverted the fix, and the bug can be seen in different EXPLAIN variants, perhaps the easiest is EXPLAIN (OPT):
```
                  │    └── inner-join (lookup array_tab [as=a1])
                  │         ├── lookup columns are key
                  │         ├── inner-join (inverted array_tab@foo_inv,inverted [as=a1])
                  │         │    ├── flags: force inverted join (into right side)
                  │         │    ├── inverted-expr
                  │         │    │    └── a1.b <@ a2.b
                  │         │    ├── scan array_tab_not_idx [as=a2]
                  │         │    │    ├── constraint: /1: [/2 - /6]
                  │         │    │    └── flags: force-index=array_tab_not_idx_pkey
                  │         │    └── filters (true)
                  │         └── filters
                  │              └── a2.b && a1.b
```
Note that the inverted expr is incorrectly replaced with ContainedBy, but then, after we perform lookup join into the PK of `array_tab`, we apply the correct filter (because that filter wasn't replaced in `extractJSONOrArrayJoinCondition`). It just so happened that this bug had no actual effect on the output :smiley:
</blockquote></details>

TIL, thanks for the detailed explanation!!


<!-- Sent from Reviewable.io -->

ZhouXing19 commented:
TFTR!
bors r+

mgartner commented:
Great work @ZhouXing19 and great review @yuzefovich! That was an interesting "non-bug" you both discovered with the overlap operator!
